### ISSUE TITLE
AssertionError when indexing with a dotted attribute and various other cases.

### ISSUE THREAD

**Issue by brianm78** [xonsh/xonsh#4079]
Attempting to index into a list or dictionary using an attribute access inside the square brackets leads to an AssertionError in the parser.  Eg:

    d = {}
    d[x.value] = 20

Gives me:

    Traceback (most recent call last):
      File "/usr/lib/python3.9/site-packages/xonsh/ptk_shell/shell.py", line 322, in _push
        code = self.execer.compile(src, mode="single", glbs=self.ctx, locs=None)
      File "/usr/lib/python3.9/site-packages/xonsh/execer.py", line 135, in compile
        tree = self.parse(input, ctx, mode=mode, filename=filename, transform=transform)
      File "/usr/lib/python3.9/site-packages/xonsh/execer.py", line 95, in parse
        tree, input = self._parse_ctx_free(input, mode=mode, filename=filename)
      File "/usr/lib/python3.9/site-packages/xonsh/execer.py", line 219, in _parse_ctx_free
        tree = self.parser.parse(
      File "/usr/lib/python3.9/site-packages/xonsh/parsers/base.py", line 497, in parse
        tree = self.parser.parse(input=s, lexer=self.lexer, debug=debug_level)
      File "/usr/lib/python3.9/site-packages/xonsh/ply/ply/yacc.py", line 335, in parse
        return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
      File "/usr/lib/python3.9/site-packages/xonsh/ply/ply/yacc.py", line 1122, in parseopt_notrack
        p.callable(pslice)
      File "/usr/lib/python3.9/site-packages/xonsh/parsers/base.py", line 2253, in p_atom_expr
        p[0] = self.apply_trailers(p[1], p[2])
      File "/usr/lib/python3.9/site-packages/xonsh/parsers/base.py", line 2247, in apply_trailers
        assert False
    AssertionError

Splitting it into 2 steps works.  Ie:

    y = x.value
    d[y] = 20

Looking a the code, it seems to be expecing trailer to be one of ast.Index, ast.Slice, ast.ExtSlice, ast.Constant or ast.Name, or else a Mapping,tuple,  string, tuple, or string, with an assert False if none match.

But it seems to be an ast.Attribute in this case.

Trying a few other cases, it looks like this happens for almost anything more complex than a simple literal or single variable.  Eg, the below all give the same issue:

    d[f()]     # trailer is an ast.Call
    d[x[y]]    # ast.Subscript
    d[1+2]     # ast.BinOp

This happens both at the prompt, and inside a xonchrc file.

**Comment by anki-code** [comment]
#3588

**Comment by anki-code** [comment]
Fixed in master by #4101.


Tested in docker:
```python
docker run --rm -it python:3.9-slim /bin/bash -c "apt update && apt install -y git && pip install -U git+https://github.com/xonsh/xonsh.git && xonsh"
```
